pacsman/pynetdicom_client.py

```diff
@@ -0,0 +1,399 @@
+
+logger = logging.getLogger(__name__)
+
+# http://dicom.nema.org/dicom/2013/output/chtml/part07/chapter_C.html
+status_success_or_pending = [0x0000, 0xFF00, 0xFF01]
+
+
+class PynetdicomClient(DicomInterface):
+
+    def verify(self):
+        """
+        :return: True on success, False on failure
+        """
+        ae = AE(ae_title=self.client_ae, scu_sop_class=['1.2.840.10008.1.1'])
+        # setting timeout here doesn't appear to have any effect
+        ae.network_timeout = self.timeout
+
+        with association(ae, self.pacs_url, self.pacs_port) as assoc:
+            logger.debug('Association accepted by the peer')
+            # Send a DIMSE C-ECHO request to the peer
+            # status is a pydicom Dataset object with (at a minimum) a
+            # (0000, 0900) Status element
+            status = assoc.send_c_echo()
+
+            # Output the response from the peer
+            if status.Status in status_success_or_pending:
+                logger.debug('C-ECHO Response: 0x{0:04x}'.format(status.Status))
+                return True
+            else:
+                logger.warning('C-ECHO Failure Response: 0x{0:04x}'.format(status.Status))
+                return False
+
+        return False
+
+    def search_patients(self, search_query):
+        """
+        Uses C-FIND to get patients matching the input (one req for id, one for name)
+        :param patient_name_search: Search string for either patient name or ID
+        :return: List of PatientInfo
+        """
+        ae = AE(ae_title=self.client_ae, scu_sop_class=QueryRetrieveSOPClassList)
+
+        with association(ae, self.pacs_url, self.pacs_port) as assoc:
+            # perform first search on patient ID
+            id_responses = _call_c_find_patients(assoc, 'PatientID',
+                                                 f'*{search_query}*')
+            # perform second search on patient name
+            name_responses = _call_c_find_patients(assoc, 'PatientName',
+                                                   f'*{search_query}*')
+
+            uid_to_result = {}
+            for (status, result) in chain(id_responses, name_responses):
+                logger.debug(status)
+                logger.debug(result)
+                if status.Status not in status_success_or_pending:
+                    raise Exception('Patient C-FIND Failure Response: 0x{0:04x}'.format(status.Status))
+                if result:
+                    # remove non-unique Study UIDs
+                    #  (some dupes are returned, especially for ID search)
+                    uid_to_result[result.StudyInstanceUID] = result
+
+            # separate by patient ID, count studies and get most recent
+            patient_id_to_info = {}
+            for study in uid_to_result.values():
+                patient_id = study.PatientID
+                study_id = study.StudyInstanceUID
+                if id in patient_id_to_info:
```

Comment:
Where is `id` being set?

Reply:
Good catch, this is fixed in latest along with using `num_studies` instead of `study_ids` below.